### PR TITLE
Removes the camelCasing of filter names - 

### PR DESCRIPTION
**Pull request by tabacitu**
Thereby fixing the issue #3065 regarding filters. Basically reverts to the way things were before #2872 regarding filter names, removing the breaking change introduced by it where filter names were unexpectedly camelCased.

After this PR, the filters names are still consistent across the filter files, BUT they're also the exact same name that the developer passed to `addField()`, no camelCasing or anything.

Please PLEASE help me test this thing, to make sure it's a good fix, or at least fixes the backwards compatibility issues 😁 

Thanks a lot!

### CODE REVIEW

**Comment by tabacitu** [comment]
Then again, wouldn't this constitute a breaking change for people who have been developing/upgrading their apps since PR 
#2872 was merged (May 25th)? That's a lot of time, I'm wondering why we're only hearing about it now...

**Comment by trip-somers** [comment]
Tough call on whether or not it's a breaking change. For it to break someone's code, someone would have to (1) be using a non-camelCase name for the filter _and_ (2) be expecting the camelCase version of that name somewhere else. It's possible that someone has done this, but even in the patch I applied to my implementation, I went ahead and changed the filter name to match the camelCase version.

As for why it wasn't brought up before, my best guess is that not a lot of people have written chained filters with names that get transformed by the camelCasing. For example, if I had used 'manufacturer', 'line', and 'series' without the '_id' suffix that I chose to use, I never would have encountered the problem. (Also, May 25th wasn't _that_ long ago for people who are already happily using a prior version and haven't considered upgrading.)

I will apply this pull request to verify that it solves my backwards compatibility issue then report back.

**Comment by trip-somers** [comment]
I have confirmed that removing those 6 lines solves the unexpected input field name problem.

**Comment by pxpm** [comment]
Hello @tabacitu and @trip-somers 

First of all thank you @trip-somers for great issue report.

I have been thinking about this, I do agree with you that if this is a breaking change (when merging we did not think that this could be one) it should have been documented. For that I am sorry. 

About the change I stand by it and why it was made.

Before the change filters had `Str::slug`, `Str::camel` all messed up in files and used 'per need', but that lead to #2829 and that was the principle why this change was added. 

We are not forcing the developer on name choosing, it can be `my_filter`, `my filter`, `my-filter`. The thing is that this naming freedom does not play well with our current implementation of filters.

Sometimes you need the filter name to build an url, other times to declare a javascript function (that can't contain ` - ` for example).

This change was made to clean up filter files, conventionate `Str::camel` and fix a bug.

I am not against conventions, Laravel uses them alot, and for me this seems a simple solution to have for people that use the filter name for some after process like you. For the people that just add the filters to their CRUD's, the naming part will not break the filters anymore, because `Str::camel` makes it work for urls, javascript functions or HTML attributes.

So I stand by adding this to 4.1 docs as a breaking change.

- I don't see many people using an undocumented functionality.
- I think the renaming process is not hard work in general (There can be one project where it would be impossible to rename because x y z ..)

If I did not made myself clear please ask anything you feel. 

Wish you guys the best,
Pedro

**Comment by trip-somers** [comment]
Why are names like 'my-field' and 'my field' being accommodated? I can't come up with a scenario in which a developer uses names like that.

**Comment by pxpm** [comment]
@trip-somers sorry I think that must be some misunderstanding here, we are talking about filters not fields. 

There is always another option, that would be not to change filter name, but validate filter name, so it can't contain spaces and dashes that don't work as javascript function name or as an url parameter. 

From my understanding it seems less restrictive for people that only want to add a filter to a crud, and I think here lays the majority of the users, but I might be wrong, to write the name they want and the filter works because we really don't care about the name, we just want it to work how it's expected. 

Altough I agree that renaming would not be the best for people who want to use advanced functionality of the filters, me personaly if using filter names just write them `camelCased`.

To avoid renaming or restricting when the name does not really care, we could introduce some filter flag like: `advanced => true` or `developer => true` that would not rename filters but expect developer to use a proper name that works with the intended filter. 

Let me know,
Pedro

**Comment by trip-somers** [comment]
Let me reset my thoughts. While I *do* think it's weird that filter names are changed, it is an easy enough thing to deal with if it is clearly documented that the actual input name will be wind up in `camelCase`.

I would definitely be in favor of something that allows me to say definitively, "This is my field name." As for what that option is called, I think something fairly explicit like an extra option called `input` (or similar) to hold the intended name would be better than something implicit like setting an option called `developer` to `true`.

**Comment by tabacitu** [comment]
Update: just talked about this with @pxpm and decided the best way forward would be to:

**(1) Push a fix ASAP, that gives back the developers the ability to use _whetever they want_ as the filter name. And have the same name back as the URL parameter.** Since this inconvenience was introduced as a non-breaking change, it should be ok to push the fix as a non-breaking-change too. We will only upset @trip-somers (sorry man) and a few others who already know about it, and have to change _back_ their names. But the alternative is worse - that people update Backpack, get this unexpected change, and more people have to do this fix inside their code... Bleah.

**(2) In addition to the current PR, have a second attribute on the filter, named `$filter->key`.** That attribute can be a camelCased name, no problem, and we can use `$filter->key` throughout filter blade files instead of `$filter->name` whenever we need to make sure that it doesn't have disallowed characters. So we use an unchanged `name` for the filter name. A changed key for everything else, if needed. @pxpm said he's on this, he'll modify this PR or send a different PR to do that.

Thanks a lot @trip-somers for confirming this PR fixes it for you, and for all the help... and for being so patient with this 😄  We messed up, but I think by the end of the week we'll have it fixed, and we can finally forget about it 🥳 

Cheers!

**Comment by trip-somers** [comment]
Happy to see this fix, and it doesn't bother me at all.

I changed my filter names to camelCase to match the expected transformation, so this won't even break my stuff. I'll just have a small inconsistency where my filter names will be camelCase while all of my other fields are snake_case, which is not a big deal and be switched back to snake_case at any time once this fix is in.

**Comment by pxpm** [comment]
@tabacitu just like we discussed, I introduced the `key` property in filters that would be a camelCased version of name. 

We use that key in internal ids, css and js. We use the name as the `url parameter` unchanged.

**Comment by tabacitu** [comment]
Excellent work here @pxpm ! Tested, it works perfectly now, thanks a lot! Merging into `master`. It'll be available officially later today after I tag it as 4.1.19

@trip-somers please remember to _change back_ your names to snake_case if you want. Thank you for being so patient, and sorry again for having pushed this bug mid-release-cycle. 

# TLDR

*What happened:* In Backpack 4.1 we mistakenly pushed an update as non-breaking when it was breaking. That update converted filter names from their usual string to `camelCase`. This was done in order to clean up the blade file code. But it has an unexpected effect - some developers actually used those filter names for other purposes too, like complicated logic where filters are dependent on one another.

*The solution:* This PR brings back the old functionality, the *filter names are no longer camelCased*, they are kept exactly the way the developer added them. Starting with Backpack 4.1.19 you can expect your filter names to be intact.

**Comment by ziming** [comment]
Yup, I changed my filter name to camelCase. Thanks for informing me about the change back to snake_case :)

**Comment by maxacarvalho** [comment]
@tabacitu and @pxpm  

First, I'd like to thank you very very much for making my team and I spent 4 hours trying to understand what the hell was going on since our filters stopped working, again!  
First your changes had our application broken since we couldn't understand what was going on with the filters, then you release another change, to revert the initial one without a fallback 👏 👏 👏 👏 👏 👏  **I AM TRILLED**

Now, I'd like to ask you to use some type of communication about this releases of yours, send emails, scream about it!  
You have the emails of everybody that bought your product, yes, **bought** your product, because this is a paid thing, so I'd expect some respect.

I know that your product is good and saved hours of development for many developers across the world but that does not take the blame out, you broke out application twice.  
I opened the first issue, reporting the first time you broke our platform. Then, from that issue, I was asked about the rollback and I said that would be best to have a backward compatibility in place, it seems that it was completely ignored.

**Comment by tabacitu** [comment]
Hi @maxacarvalho ,

I'm sorry to hear about the frustration we've caused you. Of course, you have all the reason to be upset - we fucked up. We introduced a breaking change as non-breaking, then pushed the fix as a non-breaking change.

I take all the blame for this, it was my decision to "break your app" - again. I'm truly sorry for it.

Unfortunately a tough decision had to be made, because either way we were upsetting some Backpack developers. So we had to choose between:
- upsetting a few developers who already noticed the error and started using camelCase;
- upsetting a lot more developers who did not notice it _yet_, and were just now discovering the bug; people who might have jumped over the versions with the bug;

As you can see, it was a tough one, but I think it was the right call to make under the circumstances - we chose to upset the least amount of people.

I'm not trying to excuse myself, I'm just trying to provide clarity over the issue - that I was bound to upset someone, and... I chose to upset you. As nasty as that sounds 🙄

----

I think you're right about communication. We could have done a better job of communicating this. 

I chose not to send an email blast to everyone in our "_security and major releases_" list because... it's not a security issue, or a major release. I never even considered it, since we only send an email there once/twice a year - we try to not bother people at all. Breaching this trust and spamming thousands and thousands of people would have been bad too.

You're right though - I now realize that it does make sense to have a communication channel where we can post updates more frequently. We couldn't send a blast to _that list_ but we should have _another list_.

So here's what we'll do:
- we'll start posting weekly/monthly updates on BackpackForLaravel.com as things happen that developers would like to know;
- we'll add an email digest to it;
- we'll add a RSS feed to it;
- (maybe) we'll create a twitter account that will also post those;

It's unlikely this will happen again, but if it does - we can at least do a better job at communicating it - you're right. 

Hopefully the channels above will provide a lot of people with an option they're confortable with, to receive updates about Backpack more often. And us with a way to reach out in cases like this.

Again, I'm sorry, I know words won't give you back the time you've wasted on this, so please reach out next time you purchase something from us and I'll be sure to give you a discount as an extra apology.

Cheers!
